### PR TITLE
Bool/Enum scans

### PR DESCRIPTION
**Pull request by pmldrmota**
Simple extension using the existing `ListGenerator`.
The same could be achieved by swapping the `BoolParam` for an `IntParam` and scanning the values `0,1`.
The plotting could be improved by labelling ticks `True`, `False`, but I expect this type of scan to be performed mainly for comparisons along the lines of #153.

### CODE REVIEW

File: ndscan/experiment/scan_runner.py [hunk 1]
```diff
@@ -435,6 +438,9 @@ def _point_completed(self):
         # the axis coordinates.
         values = self._current_chunk.pop(0)
         for value, sink in zip(values, self._axis_sinks):
+            if isinstance(value, np.bool_):
+                # Convert as `np.bool_` is not PYON-serializable.
+                value = bool(value)
```
Comment (dnadlinger) [inline ndscan/experiment/scan_runner.py:443]: What's going on here (and above)? Why are these `np.bool_`s in the first place?

Comment (pmldrmota) [reply ndscan/experiment/scan_runner.py:443]: Probably because of `ListGenerator.points_for_level` (line 209).

Comment (pmldrmota) [reply ndscan/experiment/scan_runner.py:443]: `LinearGenerator` returns `.tolist()` - should probably just do that in `ListGenerator`, too.

**Comment by pmldrmota** [comment]
I've implemented both `bool` and `enum`-type scans, such that categorical axis labels are placed in plots.
In principle, the categorical labels could be applied to result channels as well by adding the `"categories"` key to the `spec`.
`Enum` scans don't refer to `enum.Enum` on the experiment-side yet, but I expect the `argument_editor` and `plots` implementation to still be valid even then.
At the moment, `EnumParam` behaves like `EnumerationValue` in ARTIQ, i.e. just taking on string values.

**Comment by pmldrmota** [comment]
This version enables `enum.Enum` to be used for `EnumParam`s. For every `enum`, the parameter, the handle, and the parameter store need to be created. for this, a factory function is provided.

```python
@unique
class MyEnum(Enum):
    first = auto()
    second = auto()
    third = auto()

(MyEnumParam, MyEnumParamHandle, MyEnumParamStore) = enum_param_factory(MyEnum)
```

If the `Enum` values are strings, those strings will be used as axis tick labels and as options displayed in the argument interface.
In all other places, the `enum.name` is used to identify an option (in order to coerce it into a member using `MyEnum[name]`): when serialising the selection/scan from the UI and when determining the default value from a dataset entry. In the above example, the dataset default could be specified as `"dataset('baz', 'third')"` or as `"'third'"`, or as `MyEnum.third`.

Here is an example of scanning a `BoolParam` and a custom `EnumParam`,
```python
@unique
class MyEnum(Enum):
    first = "First option"
    second = "Second option"
    third = "Third option"
```
image

image

**Comment by pmldrmota** [comment]
NB I haven't tested this for on-kernel scans yet.

**Comment by dnadlinger** [comment]
I have a few small naming nits etc. as well, but one main issue is that we need a better solution for the type lookup. `id(…)` is not stable across time, e.g. different Python invocations such as repository scanning vs actual executation. Either this doesn't work right now, or it does, but then the `id()` output shouldn't be necessary in the first place. This may require rearchitecting ndscan's parameter store initialisation internals a bit.

I've put a rebased version here to avoid duplication of work, but didn't get to actually tackle this today: https://github.com/OxfordIonTrapGroup/ndscan/tree/wip-enum-scans

**Comment by dnadlinger** [comment]
I have on-device scans working, but it requires an ARTIQ compiler fix to be able to handle scans of multiple different `EnumParam` types.

**Comment by dnadlinger** [comment]
Merged in https://github.com/OxfordIonTrapGroup/ndscan/pull/399.

**Pull request closed without merging**